src/main/java/com/autotune/analyzer/recommendations/engine/DurationBasedRecommendationEngine.java

```diff
@@ -379,6 +379,10 @@ private static RecommendationConfigItem getCPURequestRecommendation(Map<Timestam
             cpuRequest = CommonUtils.percentile(NINETY_EIGHTH_PERCENTILE, cpuUsageList);
         }
 
+        // Make sure that the recommendation cannot be 0 or -ve
```

Comment:
If the cpu usage is zero at this point, that means that the usage for atleast 24 hrs (short term) was zero. I'd say that we should create a notification at this point that the (appropriate) term has not enough cpu data to arrive at a recommendation.

Also, depending on which env the workload in running, `0.1` can be too much! (Eg embedded), so not good to fix that. Unless there is no history of any usage, what this means is that the workload was idling during this observation period and that should just be a notification saying the same

Reply:
@dinogun This gets me a thought on the flow of the recommendations, I was in an impression that if we have data with us we need to provide a recommendation which is satisfactory, So every result after a certain threshold (min data) we generate a recommendation. Here are my views on how we see this

**Assumption-A: Every result after a certain min data requirement should have a recommendation**

In this type of assumption, we check if the recommendation is non zero, if it's zero then we lookup for last applied cpuRequest and recommend the same, if there is no cpuRequest set we return min recommendation hardcoded.

**Assumption-B: Recommendation is provided only if some constraints are met even after min data requirement**

In this type of assumption, we check if the recommendation is non zero, if it's zero we add a notification saying that recommendation cannot be generated as the workload is idle

I was thinking if we extend assumption A a bit further, we can detect the idle entries (cpu usage = 0) and avoid them for recommendation calculation and calculate recommendation on only non-idle data, This makes us consider the 0 cpuUsage entries as missing data and we will not consider that entries duration in minutes at all, So this keeps it straight forward that if there is min valid and non - zero data for a term (eg. 24 hrs) exists before we give our first recommendation, else we by default add not enough data for recommendation notification.

I'm interested to know in which way we can go forward Assumption A or B. I couldn't decide upon one way as each of them has their own share of pros and cons. Please lemme know your views on this.

PS: I have added my views (in Assumption A) assuming that having notification after min data is mandatory, if my base assumption itself is wrong (please correct me if it is) then we can proceed with Assumption B.

Thanks in advance!

Reply:
@bharathappali  By generating recommendations based on available data alone, both relationship building and advanced analysis can be achieved. For example, identifying consumption patterns for weekday,  specific events or holidays.
Notifications can also inform users about these behavioral changes, and if needed, they can consider previous or short-term recommendations.

In cases where a memory-bound pod is present, CPU consumption may drop to zero while memory usage remains high. So, it is important not to ignore idle CPU usage.
So, it would be better to consider the data available during that duration alone and provide recommendations or notifications.